File: breadth_first_search/breadth_first_search.py
graph_example = {
    "N": ["S", "NO", "NE"],
    "S": ["N", "SO", "SE"],
    "O": ["L", "NO", "SO"],
    "L": ["O", "NE", "SE"],
    "NO": ["N", "O"],
    "NE": ["N", "L"],
    "SO": ["S", "O"],
    "SE": ["S", "L"],
}


from collections import deque

def bfs(vertex: str, queue: deque, visited: list, graph: dict) -> None:
    if vertex not in visited:
        visited.append(vertex)
    
    neighborhood = graph[vertex]

    for neighbor in neighborhood:
        if neighbor not in visited:
            queue.append(neighbor)
    
    if queue:
        next_vertex = queue.popleft()
        if next_vertex:
            bfs(next_vertex, queue, visited, graph)

my_queue = deque()
result = list()

bfs("N", my_queue, result, graph_example)

print(result)

File: breadth_first_search/breadth_first_search.go
package main

import (
	"fmt"
	"slices"
)

func bfs(vertex string, queue *[]string, visited *[]string, graph map[string][]string) {
	if !slices.Contains(*visited, vertex) {
		*visited = append(*visited, vertex)
	}

	neighborhood := graph[vertex]

	for _, neighbor := range neighborhood {
		if !slices.Contains(*visited, neighbor) && !slices.Contains(*queue, neighbor) {
			*queue = append(*queue, neighbor)
		}
	}

	if len(*queue) > 0 {
		nextVertex := (*queue)[0]
		*queue = (*queue)[1:]

		bfs(nextVertex, queue, visited, graph)
	}
}

func main() {
	graphExample := map[string][]string{
		"N":  {"S", "NO", "NE"},
		"S":  {"N", "SO", "SE"},
		"O":  {"L", "NO", "SO"},
		"L":  {"O", "NE", "SE"},
		"NO": {"N", "O"},
		"NE": {"N", "L"},
		"SO": {"S", "O"},
		"SE": {"S", "L"},
	}

	myQueue := []string{}
	result := []string{}

	bfs("N", &myQueue, &result, graphExample)

	fmt.Println(result)
}
